### PR TITLE
Fixes to build-js.sh and JS API

### PR DESCRIPTION
**Pull request by rsms**
- Removes hard-coded path to Emscripten in build-js.sh in favor for `EMSCRIPTEN` environment var
- Removes invalid `ModuleInstance::` prefix from WebIDL definition of LiteralList
- Adds BufferWithRandomAccess and WasmBinaryWriter to WebIDL definition, enabling use of the JS library to output WASM module code (e.g. to be used with `Wasm.instantiateModule` or other applications requiring the actual WASM module code as byte-array data.)
- Adds some missing and additional patterns to .gitignore
- Adds -g flag to build-js.sh that will build unoptimized and unmangled versions of binaryen.js and wasm.js (output has a "-g.js" suffix to allow co-existence with regular optimized builds).
- Enables closure compiler for non-debug builds, greatly reducing file size
- Adds browser test for s-expression parser + WASM code gen (requires wasm to be enabled in the browser.) (`test/binaryen.js/browser.html`)
- Adds iterator interface to `BufferWithRandomAccess` (when `Symbol.iterator` is available)
- Adds `toArrayBuffer` to `BufferWithRandomAccess` (when `TypedArray` is available)
- Adds `compileWast(sourceText :string) :ArrayBuffer` (`compileWast(sourceText :string) :Buffer` if env is NodeJS) to the module, parsing & compiling s-expression code to a WASM module.
- Changes the way binaryen.js is exported to allow usage in CommonJS, AMD and UMD environments.

### CODE REVIEW

File: src/wasm-binary.h [hunk 1]
```diff
@@ -473,6 +473,9 @@ class WasmBinaryWriter : public Visitor<WasmBinaryWriter, void> {
   WasmBinaryWriter(Module* input, BufferWithRandomAccess& o, bool debug) : wasm(input), o(o), debug(debug) {
     prepare();
   }
+  WasmBinaryWriter(Module& input, BufferWithRandomAccess& o, bool debug) : wasm(&input), o(o), debug(debug) {
```
Comment (rsms) [inline src/wasm-binary.h]: A constructor accepting a reference instead of a pointer is required for WebIDL use (in src/js/binaryen.idl)

**Comment by kripken** [comment]
Thanks!

In `build-js.sh`, how about using the env var `$EMSCRIPTEN`? Then people could write `emmake build-js.sh` which is consistent with general emscripten build instructions (similar to `emmake make`). I'm also ok with supporting both a commandline parameter and the env var.

**Comment by rsms** [comment]
Good idea! I'll make the change @kripken

**Comment by rsms** [comment]
Okay, no more command-line argument, but instead uses `EMSCRIPTEN` env var

File: src/wasm-binary.h (hunk 1 above)
Comment (kripken) [inline src/wasm-binary.h]: Ok, this PR looks great now, except that I think this is unnecessary - WebIDL should be able to handle a simple pointer? In fact that's the default behavior, so just removing the `[Ref]` should make it work.

**Comment by rsms** [comment]
PR has been updated with additional changes

**Comment by kripken** [comment]
Something seems to have gone wrong here, travis reports the binaryen.js test now fails.

**Comment by rsms** [comment]
Managed to repro. a.js generated by check.py seems to simply concatenate binaryen.js instead of loading it separately (relied on global "this" to reflect window, global, etc.) Changes makes export explicit to `window || global || this` (check.py related to binaryen.js passes for me.)

**Comment by rsms** [comment]
FYI: Another change is incoming...

**Comment by rsms** [comment]
Still failing, but when I run `check.py` locally (macOS 11) it all runs OK:

On my machine:
```
...
[ checking binaryen.js testcases... ]

test.js
executing:  node a.js

[ checking .s testcases... ]

.. alias.s
executing:  bin/s2wasm test/dot_s/alias.s --emscripten-glue
...
```

Appveyor (https://travis-ci.org/WebAssembly/binaryen/jobs/155457595#L2979):

```
...
/home/travis/build/WebAssembly/binaryen/a.js:28
cess.on("uncaughtException",function(a){if(!(a instanceof ta))throw a;});d.ins
                                                                    ^
ReferenceError: Binaryen is not defined
    at Object.<anonymous> (/home/travis/build/WebAssembly/binaryen/a.js:437:12)
    at Module._compile (module.js:456:26)
    at Object.Module._extensions..js (module.js:474:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:312:12)
    at Function.Module.runMain (module.js:497:10)
    at startup (node.js:119:16)
    at node.js:929:3
Traceback (most recent call last):
  File "./check.py", line 475, in <module>
    out = run_command(cmd)
  File "/home/travis/build/WebAssembly/binaryen/scripts/support.py", line 145, in run_command
    raise Exception(('run_command failed', err))
Exception: ('run_command failed', None)
The command "./check.py" exited with 1.
```

**Comment by rsms** [comment]
Okay! Tests passed.

You can try opening `test/binaryen.js/browser.html` in FF or Chrome. It essentially loads binaryen.js and does this, printing info to `body` as it goes:

```js
let buf = Binaryen.compileWast(input1);
Wasm.verifyModule(buf);
let m = Wasm.instantiateModule(buf, {
  env: {
    add: function(a, b) { return a + b; }
  }
});
let res = m.exports.add(10.0, 20)
```

File: test/binaryen.js/browser.html [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+    </style>
+    <script type="text/javascript" src="../../bin/binaryen.js"></script>
+    <!-- <script type="text/javascript" src="../../bin/binaryen-g.js"></script> -->
+  </head>
+  <body><script type="text/javascript">
+
+if (!window.Wasm) {
+  document.write('No WASM support. Did you enable it via chrome://flags/#enable-webassembly ?')
```
Comment (kripken) [inline test/binaryen.js/browser.html]: please link to https://github.com/kripken/emscripten/wiki/WebAssembly#testing-native-webassembly-in-browsers which has those chrome instructions plus others.

File: src/js/binaryen.js-post.js [hunk 3]
```diff
@@ -0,0 +1,29 @@
+return Module;
```
Comment (kripken) [inline src/js/binaryen.js-post.js]: Instead of  the `pre

Comment (kripken) [inline src/js/binaryen.js-post.js]: Instead of the `pre` and `post` bits to wrap and return the module, you can use emcc's `MODULARIZE` option. although then the code outside the function must be added after `emcc`, so maybe not worth it.

File: src/js/binaryen.js-extended.js [hunk 4]
```diff
@@ -0,0 +1,56 @@
+// Extended API
+
+// Makes the internal buffer iterable so we can use Uint8Array.from
+if (typeof Symbol !== 'undefined' && Symbol.iterator) {
```
Comment (kripken) [inline src/js/binaryen.js-extended.js]: i'm having a hard time understanding this... what is `Symbol`? what does the code do with it?

File: src/js/binaryen.js-post.js (hunk 3 above)
Comment (rsms) [inline src/js/binaryen.js-post.js]: I attempted this but was unable to make it work with closure compiler and commonjs/amd environments.

File: src/js/binaryen.js-extended.js (hunk 4 above)
Comment (rsms) [inline src/js/binaryen.js-extended.js]: The [iterable protocol](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols#iterable) relies on `Symbol.iterator` and thus for a browser without `Symbol` (ES6 feature), the `BufferWithRandomAccess` is not extended to be iterable.

`[Symbol.iterator]` is how the iterable protocol is implemented. Making `BufferWithRandomAccess` iterable allows using it with `TypedArray.from(iterable)`. There might be a better way to do this on a higher lever, by using TypedArray in some form (wrapped?) in the Binaryen code, avoiding using the internal `BufferWithRandomAccess`. But for now I think this is a good start.

**Comment by rsms** [comment]
Screenshot of what `test/binaryen.js/browser.html` looks like when `window.Wasm` is undefined:

screenshot 2016-08-30 15 33 06

File: build-js.sh [hunk 5]
```diff
@@ -3,12 +3,113 @@
...
+EMCC_ARGS="$EMCC_ARGS -s DEMANGLE_SUPPORT=1"
+OUT_FILE_SUFFIX=
+
+if [ "$1" == "-g" ]; then
+  EMCC_ARGS="$EMCC_ARGS -O0"
+  EMCC_ARGS="$EMCC_ARGS --llvm-opts 0 --llvm-lto 0"
+  EMCC_ARGS="$EMCC_ARGS -profiling"
+  OUT_FILE_SUFFIX=-g
+else
+  EMCC_ARGS="$EMCC_ARGS -Oz"
```
Comment (kripken) [inline build-js.sh]: these options look different than before. just curious how you picked them? was there a specific benefit to each over plain `-Oz`?

**Comment by kripken** [comment]
Looks great, I had just one last question about the emcc flags.

Also, please join the [wasm group](https://www.w3.org/community/webassembly/participants) before we merge.

File: build-js.sh [hunk 6]
```diff
@@ -3,12 +3,113 @@
...
+  exit 1
+fi
+
+if [ "$EMSCRIPTEN" == "" ]; then
+  echo "$0: EMSCRIPTEN environment variable is not set" >&2
+  exit 1
+fi
+
+if [ ! -d "$EMSCRIPTEN" ]; then
+  echo "$0: $EMSCRIPTEN is not a directory" >&2
```
Comment (rsms) [inline build-js.sh]: This is wrong. `$EMSCRIPTEN` will be expanded in the message to the empty string. I'll change this to `\$EMSCRIPTEN`

File: build-js.sh (hunk 5 above)
Comment (rsms) [inline build-js.sh]: @kripken Disabling optimizations for the debug build makes it easier to look through the generated code and to draw parallels to the source IMHO. Happy to change it to `-Oz` for both targets if you prefer. Let me know!

Comment (kripken) [inline build-js.sh]: That part makes sense, yeah. I'm more curious about the llvm opts, lto, and duplicate function stuff.

**Comment by rsms** [comment]
https://www.w3.org/users/93385

File: build-js.sh (hunk 5 above)
Comment (rsms) [inline build-js.sh]: When building Figma to asm.js, we've found these settings to generate small and fast code.

However, I'm right now running a bunch of variations of the flags for binaryen.js and will post the results with file size and perf measurements here in a few minutes.

**Comment by rsms** [comment]
See https://gist.github.com/rsms/e33c61a25a31c08260161a087be03169

TL;DR: `-Oz --llvm-lto 1 -s ELIMINATE_DUPLICATE_FUNCTIONS=1` yields best over-all performance in my measurements.

**Comment by rsms** [comment]
Travis build error seems to be unrelated to the JS version:

```
...
executing:  bin/asm2wasm test/hello_world.asm.js
No output has been received in the last 10 minutes, this potentially indicates a stalled build or something wrong with the build itself.
The build has been terminated
```

**Comment by kripken** [comment]
Nice data. Yeah, that's fairly consistent with other projects I've seen. I'm always surprised how useful duplicate function elimination is, but it seems to help on any large-enough C++ codebase.

Btw, does `build-js.sh` work if `$EMSCRIPTEN` is not defined? It might be nice to not fail in that case if `em++` is in the path already.

**Comment by kripken** [comment]
Thanks! Is this ready to land?

**Comment by rsms** [comment]
Yup! 😃

**Comment by kripken** [comment]
Great stuff, thanks!
